### PR TITLE
Join enums refactoring

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Extracted join related enumerations from ASTTableJoin into separate file in Core.
Necessary to remove unnecessary dependencies between Parsers, Analyzer, Interpreter. Related to https://github.com/ClickHouse/ClickHouse/pull/31796.

### CODE REVIEW

File: src/Core/Joins.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#include <Core/Joins.h>
+
+namespace DB
+{
+
+const char * toString(JoinKind kind)
```
Comment (vdimir) [inline src/Core/Joins.cpp:6]: Is `magic_enum` intentionally not used?

File: src/Core/Joins.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+namespace DB
+{
+
+/// Join method.
```
Comment (vdimir) [inline src/Core/Joins.h:6]: Also, we have this enum used to improve readability in some cases.

https://github.com/ClickHouse/ClickHouse/blob/24aaac59ca207fd3dcd1bea34a67acc86fba58d0/src/Interpreters/TableJoin.h#L45-L49

**Review by vdimir — APPROVED** [review]
👍

File: src/Core/Joins.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Core/Joins.cpp:6]: Yes. It had some issues with lifetime of enum entry. We have workaround in ASTSystemQuery.cpp, but here it is simpler to just manually convert to string.

File: src/Core/Joins.h (hunk 2 above)
Comment (kitaisreal) [reply src/Core/Joins.h:6]: Updated.

**Pull request merged**